include/ppx/metrics.h

```diff
@@ -215,11 +248,27 @@ class Manager final
 public:
     Manager() = default;
 
-    Run* AddRun(const std::string& name);
+    // Starts a run. There may only be one active run at a time.
+    void StartRun(const std::string& name);
+    // Concludes the current run.
+    void EndRun();
+    // Returns whether a run is active.
+    bool HasActiveRun() const;
+
+    // Adds a metric to the current run. A run must be started to add a metric.
+    // Failure to add a metric returns kInvalidMetricID.
+    MetricID AddMetric(const MetricMetadata& metadata);
+
+    // Record data for the given metric ID. Metrics for completed runs will be discarded.
+    bool RecordMetricData(MetricID id, const MetricData& data);
 
-    // Exports all the runs and metrics information into a report to disk.
+    // Exports all the runs and metrics information into a report to disk. Does NOT close
+    // the current run.
     void ExportToDisk(const std::string& reportPath, bool overwriteExisting = false) const;
 
+    // Exports all current data to a string.
+    std::string ExportToString(const std::string& reportPath) const;
```

Comment:
Because it generates the full report, it needs the reportPath to create the first entry in the report itself (filename).

I debated whether it really was necessary to get this, but I figured it would improve the testability of the report functionality if we had it (that is, the thing we test is the ACTUAL report that will be written to disk otherwise).

On the other hand, tests now technically involve more integration because of it. So... I'm not sure which is preferable?

This exposure at the manager level IS a test-only function. We could alleviate that somewhat by exposing the report itself. We could also go even further and expose the json blob rather than have to parse the whole blob each time we run through the tests. I don't have a good sense of which is the best route. Your thoughts?

Reply:
On reflection, I think the best way to handle it will just be to expose the Report object again. I pulled it into the Manager because I was worried about data ownership, but after some experimentation it seems my fears were unfounded. That way we would:
1. Create the report, with the report path (always).
2. Get the report.
3. Ask the report to either write to disk or just get its string.

In that way, the semantics are clear.